### DISCUSSION TITLE
Use `base64enc` instead of `base64url`

### DISCUSSION THREAD

**Discussion by dipterix** [ropensci/targets#976] [Ideas]
I just realized that you could use `base64enc` instead of `base64url` to encode strings to `base64` (see [implementation](https://github.com/dipterix/dipsaus/blob/bab8b277f52d0abd0f0ad2f96503489fa69ada58/R/strings-base64.R#L1-L29)). The benefit of `base64enc` is you can encode any files/binaries to base64, not only characters. 

Use case 1: save images as `base64` strings. `base64` strings could be easily loaded as `src` attributes in html `img` tag

Use case 2: serialize R expressions (e.g.`tar_format`): instead of `base64_urlencode(deparse1(fun))`, you could 

```r
base64 <- base64enc::base64encode(serialize(fun, NULL))

unserialize(base64enc::base64decode(base64))
```

to preserve the formatter's environment to reduce unexpected behaviors.

One issue I encountered was I was trying to implement a tar_format dynamically in my package. However, I had to use 

```
pkg <- asNamespace("my package")
pkg$do_something(...)
```
Inside of my formatter to satisfy both `targets` and `CRAN` checker (as I cannot use `mypackage::do_something` in my own package.

**Comment by wlandau** [comment]
It's an interesting thought, but I deliberately chose to only store the deparsed text of a function and discard any attached environments etc. Such environments are not needed most of the time, and in the general case, they are serialized by accident cause surprises because the total object size becomes unmanageably large. In this particular case, the consequences would be particularly egregious because the output of `tar_format()` is stored in the metadata (`_targets/meta/meta`) which needs to stay lightweight or else the whole pipeline will have issues.

> ...to preserve the formatter's environment to reduce unexpected behaviors.

I believe this is not as surprising as nor as common the large data issue I mentioned. And https://docs.ropensci.org/targets/reference/tar_format.html#format-functions documents how these functions are stored:

> targets converts each function to and from text, so it must not rely on any data in the closure. This disqualifies functions produced by [Vectorize()](https://rdrr.io/r/base/Vectorize.html), for example.

**Reply by dipterix** [reply]
Gotcha, I would have done the same if I were you. I was wondering if it is feasible to allow the `format` to stay at least within package named space, or at least specify with attributes which namespace it can run under?

My proposed approach is

```r
read_fun <- function(...) { ... }
attr(read_fun, "target-format-namespace") <- "stats"
tar_format(read = read_fun, ...)
```

When unserializing formatter, the function `read` can be placed with environment being `new.env(parent = asNamespace("stats")`. Compared to completely changing the function serializing method, this approach is backward-compatible and easier to implement.

**Reply by wlandau** [reply]
I would prefer not to get directly involved in namespaces for those functions because it is usually fragile to try to modify package namespaces post-hoc, particularly base ones like stats. I suggest you write a custom function wrapper and add any `::` or `:::` statements you need inside it.

**Reply by dipterix** [reply]
Making a function's environment to be **parented** by a namespace does not change the namespace. Whatever generated within the function's environment won't affect the parent namespace, let alone the namespace are most likely to be locked. 

```r
environment(f) <- new.env(parent = asNamespace("..."))
```

I was wondering where does formatter's reader & writer unserialized? There must be an environment for the function to execute. I hope it's not `globalenv()` or some unlocked random environment.

**Reply by wlandau** [reply]
The function text gets converted back into a function using a temporary environment that inherits from base:

https://github.com/ropensci/targets/blob/0f08ac9f308dc6e4bce130240df44679cc73713b/R/class_store_custom.R#L86-L90

**Reply by dipterix** [reply]
Oh great to know! This looks robust. What I suggested was to be able to put `what` instead of inside of `new.env(parent = baseenv())`, be inside of `new.env(parent = asNamespace(ns))` (just for package developers who want to extent `targets`. Note that CRAN allows devs to use `asNamespace` in their packages.)

```r
store_custom_call_method <- function(text, args) {
  envir <- new.env(parent = baseenv())
  what <- eval(parse(text = text), envir = envir)
  ns <- attr(what, "target-format-namespace")
  if(length(ns) == 1 && nzchar(ns)) {
    environment(what) <- new.env(parent = asNamespace(ns))
  }
  do.call(what = what, args = args, envir = envir)
}
```

**Reply by wlandau** [reply]
I think I understand your proposed implementation, but I am not sure I see the advantage (as opposed to wrapping a call to `tar_format()` in your package and exporting that wrapper to users).
